kernel/src/checkpoint/mod.rs

```diff
@@ -1,8 +1,390 @@
+    ///
+    /// This method **must** be called only after:
+    /// 1. The checkpoint data iterator has been fully consumed
+    /// 2. All data has been successfully written to object storage
+    ///
+    /// # Parameters
+    /// - `engine`: Implementation of [`Engine`] apis.
+    /// - `metadata`: The metadata of the written checkpoint file
+    /// - `checkpoint_data`: The exhausted checkpoint data iterator (must be fully consumed)
+    ///
+    /// # Returns: [`variant@Ok`] if the checkpoint was successfully finalized
+    #[allow(unused)]
+    fn finalize(
+        self,
+        _engine: &dyn Engine,
+        _metadata: &FileMeta,
+        _checkpoint_data: CheckpointDataIterator,
+    ) -> DeltaResult<()> {
+        // Verify the iterator is exhausted (optional)
+        // Implementation will use checkpoint_data.actions_count and checkpoint_data.add_actions_count
+
+        // TODO(#850): Implement the actual finalization logic
+        Err(Error::checkpoint_write(
+            "Checkpoint finalization is not yet implemented",
+        ))
+    }
+
+    /// Creates the checkpoint metadata action for V2 checkpoints.
+    ///
+    /// This function generates the [`CheckpointMetadata`] action that must be included in the
+    /// V2 spec checkpoint file. This action contains metadata about the checkpoint, particularly
+    /// its version.
+    ///
+    /// # Implementation Details
+    ///
+    /// The function creates a single-row [`EngineData`] batch containing only the
+    /// version field of the [`CheckpointMetadata`] action. Future implementations will
+    /// include the additional metadata field `tags` when map support is added.
+    ///
+    /// # Returns:
+    /// A [`CheckpointBatch`] batch including the single-row [`EngineData`] batch along with
+    /// an accompanying selection vector with a single `true` value, indicating the action in
+    /// batch should be included in the checkpoint.
+    fn create_checkpoint_metadata_batch(
+        &self,
+        version: i64,
+        engine: &dyn Engine,
+    ) -> DeltaResult<CheckpointBatch> {
+        let checkpoint_metadata_batch = engine.evaluation_handler().create_one(
+            CHECKPOINT_METADATA_ACTION_SCHEMA.clone(),
+            &[Scalar::from(version)],
+        )?;
+
+        let filtered_data = FilteredEngineData {
+            data: checkpoint_metadata_batch,
+            selection_vector: vec![true], // Include the action in the checkpoint
+        };
+
+        Ok(CheckpointBatch {
+            filtered_data,
+            actions_count: 1,
+            add_actions_count: 0,
+        })
+    }
+
+    /// Calculates the cutoff timestamp for deleted file cleanup.
+    ///
+    /// This function determines the minimum timestamp before which deleted files
+    /// will be permanently removed during VACUUM operations, based on the table's
+    /// `deleted_file_retention_duration` property.
```

Comment:
can we say how it's used in the checkpoint logic (e.g. that it is for determining whether or not to include remove actions in checkpoint for _later_ VACUUM use)